core/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -228,8 +226,10 @@ public int hashCode() {
 
     /**
      * Checks for any conflicts between this field type and other.
+     * If strict is true, all properties must be equal.
+     * Otherwise, only properties which must never change in an index are checked.
      */
-    public void checkCompatibility(MappedFieldType other, List<String> conflicts) {
+    public void checkCompatibility(MappedFieldType other, List<String> conflicts, boolean strict) {
```

Comment:
I agree we need a better error message here. We can't just check the class like mappers do, since some mappers use MappedFieldType directly. I think I will add a setting to MappedFieldType which is the "typename"? This will give a much better error than what we get now IMO (the actual type, instead of a classname the user knows nothing about).